cni/cmd/install/main.go

```diff
@@ -0,0 +1,316 @@
+	serviceAccountTokenPath := serviceAccountPath + "/token"
+	serviceAccountToken, err := ioutil.ReadFile(serviceAccountTokenPath)
+	if err != nil {
+		return err
+	}
+
+	if fileExists(serviceAccountTokenPath) {
+		kubernetesServiceHost := env.GetString("KUBERNETES_SERVICE_HOST", "")
+		if kubernetesServiceHost == "" {
+			logger.Default.Error("KUBERNETES_SERVICE_HOST env variable not set")
+			os.Exit(1)
+		}
+
+		kubernetesServicePort := env.GetString("KUBERNETES_SERVICE_PORT", "")
+		if kubernetesServicePort == "" {
+			logger.Default.Error("KUBERNETES_SERVICE_PORT env variable not set")
+			os.Exit(1)
+		}
+
+		kubeCaFile := env.GetString("KUBE_CA_FILE", serviceAccountPath+"/ca.crt")
+		kubeCa, err := ioutil.ReadFile(kubeCaFile)
+		if err != nil {
+			return err
+		}
+		kubernetesServiceProtocol := env.GetString("KUBERNETES_SERVICE_PROTOCOL", "https")
+		tlsConfig := "certificate-authority-data: " + base64.StdEncoding.EncodeToString(kubeCa)
+
+		kubeconfig := kubeconfigTemplate(kubernetesServiceProtocol, kubernetesServiceHost, kubernetesServicePort, string(serviceAccountToken), tlsConfig)
+		err = atomic.WriteFile(mountedCniNetDir+"/"+kubecfgName, strings.NewReader(kubeconfig))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func kubeconfigTemplate(protocol, host, port, token, tlsConfig string) string {
+	return `
+# Kubeconfig file for kuma CNI plugin.
+apiVersion: v1
+kind: Config
+clusters:
+- name: local
+  cluster:
+    server: ` + protocol + `://` + host + `:` + port + `
+    ` + tlsConfig + `
+users:
+- name: kuma-cni
+  user:
+    token: ` + token + `
+contexts:
+- name: kuma-cni-context
+  context:
+    cluster: local
+    user: kuma-cni
+current-context: kuma-cni-context
+`
+}
+
+func copyBinaries(cniConfName string) {
+	if cniConfName == "" {
+		cniConfName = "YYY-kuma-cni.conflist"
+	}
+	dirs := []string{"/host/opt/cni/bin", "/host/secondary-bin-dir"}
+	// todo: if not written anywhere fail
+	for _, dir := range dirs {
+		if !isDirWriteable(dir) {
+			logger.Default.Warn("directory is not writeable", zap.String("dir", dir))
+			continue
+		}
+		file, err := os.Open("/opt/cni/bin/kuma-cni")
+		stat, err := os.Stat("/opt/cni/bin/kuma-cni")
+		logger.Default.Debug("/opt/cni/bin/kuma-cni file permissions", zap.Int("permissions", int(stat.Mode())))
+		if err != nil {
+			logger.Default.Error("can't open kuma-cni file", zap.Error(err))
+			os.Exit(1)
+		}
+		destination := dir + "/kuma-cni"
+		err = atomic.WriteFile(destination, file)
+		if err != nil {
+			logger.Default.Error("can't atomically write kuma-cni file", zap.Error(err))
+		}
+		err = os.Chmod(destination, stat.Mode()|0111)
+		if err != nil {
+			logger.Default.Error("can't chmod kuma-cni file", zap.Error(err))
+		}
+
+		if err != nil {
+			logger.Default.Error("can't atomically write cni file", zap.Error(err), zap.String("dir", dir))
+			os.Exit(1)
+		}
+
+		logger.Default.Info("wrote kuma CNI binaries", zap.String("dir", dir))
+	}
+}
```

Comment:
In this function, you should return error instead of logging. I don't believe `os.Exit(1)` should be placed here.